### PR TITLE
Multiple avro schema support

### PR DESCRIPTION
**Pull request by sumit-ct**
Support for deserializing avro compressed file with different schemas. 

Support for populating index and type value from the event data.

### CODE REVIEW

File: src/main/java/com/cognitree/flume/sink/elasticsearch/Constants.java [hunk 1]
```diff
@@ -70,7 +72,7 @@
     public static final String ES_CLUSTER_NAME = "cluster.name";
     public static final String DEFAULT_ES_CLUSTER_NAME = "elasticsearch";
 
-    public static final String ES_HOSTS = "es.client.hosts";
+    public static final String ES_HOSTS = "hostNames";
```
Comment (snwc) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/Constants.java]: whats the need for this change?

File: src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java [hunk 2]
```diff
@@ -85,10 +85,10 @@ public Status process() throws EventDeliveryException {
                 String body = new String(event.getBody(), Charsets.UTF_8);
                 if (!Strings.isNullOrEmpty(body)) {
                     logger.debug("start to sink event [{}].", body);
+                    XContentBuilder xContentBuilder = serializer.serialize(event);
                     String index = indexBuilder.getIndex(event);
                     String type = indexBuilder.getType(event);
                     String id = indexBuilder.getId(event);
-                    XContentBuilder xContentBuilder = serializer.serialize(event);
```
Comment (snwc) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/ElasticSearchSink.java]: Can you remind me about the purpose of this change? You did explain, but I seem to have forgotten.

File: src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java [hunk 3]
```diff
@@ -45,29 +47,46 @@
 
     private DatumReader<GenericRecord> datumReader;
 
+    private Map<String, DatumReader<GenericRecord>> avroHashSchemaToDatumReaderMap;
```
Comment (snwc) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java]: rename this field

File: src/main/java/com/cognitree/flume/sink/elasticsearch/Constants.java (hunk 1 above)
Comment (sumit-ct) [reply src/main/java/com/cognitree/flume/sink/elasticsearch/Constants.java]: In the official flume-elasticsearch-sink and elsewhere, hostname:port is taken from hostnames field.

 This change is to keep consistency with other elasticsearch sink configuration, So one can directly use this implementation without a need to change configuration.

File: src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java [hunk 4]
```diff
@@ -45,29 +47,46 @@
...
+            Decoder decoder = new DecoderFactory().binaryDecoder(event.getBody(), null);
+            GenericRecord data;
+            if(event.getHeaders().containsKey("flume.avro.schema.literal")) {
+                Schema schema = new Schema.Parser().parse((event.getHeaders().get("flume.avro.schema.literal")));
+                DatumReader<GenericRecord> datumReader = new GenericDatumReader<GenericRecord>(schema);
+                data = datumReader.read(null, decoder);
+                avroHashSchemaToDatumReaderMap.put(event.getHeaders().get("flume.avro.schema.hash"), datumReader);
             } else {
-                logger.error("Schema File is not configured");
+                if(avroHashSchemaToDatumReaderMap.containsKey(event.getHeaders().get("flume.avro.schema.hash"))) {
```
Comment (snwc) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java]: Extract constants for "flume.avro....."

File: src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java [hunk 5]
```diff
@@ -45,29 +47,46 @@
...
+
+            logger.trace("Record in event " + data);
+            XContentParser parser = XContentFactory
+                    .xContent(XContentType.JSON)
+                    .createParser(NamedXContentRegistry.EMPTY, data.toString());
+            builder = jsonBuilder().copyCurrentStructure(parser);
+
+            event.setBody(builder.string().getBytes());
+
+            parser.close();
```
Comment (snwc) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java]: should we wrap parser.close() within try-finally?

File: src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java [hunk 6]
```diff
@@ -45,29 +47,46 @@
...
-                GenericRecord data = datumReader.read(null, decoder);
-                logger.trace("Record in event " + data);
-                XContentParser parser = XContentFactory
-                        .xContent(XContentType.JSON)
-                        .createParser(NamedXContentRegistry.EMPTY, data.toString());
-                builder = jsonBuilder().copyCurrentStructure(parser);
-                parser.close();
+            Decoder decoder = new DecoderFactory().binaryDecoder(event.getBody(), null);
+            GenericRecord data;
+            if(event.getHeaders().containsKey("flume.avro.schema.literal")) {
```
Comment (snwc) [inline src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java]: pull out the logic of finding a DatumReader using the event into a separate method

File: src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java (hunk 3 above)
Comment (sumit-ct) [reply src/main/java/com/cognitree/flume/sink/elasticsearch/AvroSerializer.java]: changed to hashToReaderMap.

**Pull request merged**